text/0000-experimental-coroutines.md

```diff
@@ -0,0 +1,471 @@
+- Feature Name: `coroutines`
+- Start Date: 2017-06-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is an **experimental RFC** for adding a new feature to the language,
+coroutines (also commonly referred to as generators). This RFC is intended to be
+relatively lightweight and bikeshed free as it will be followed by a separate
+RFC in the future for stabilization of this language feature. The intention here
+is to make sure everyone's on board with the *general idea* of
+coroutines/generators being added to the Rust compiler and available for use on
+the nightly channel.
+
+# Motivation
+[motivation]: #motivation
+
+One of Rust's 2017 roadmap goals is ["Rust should be well-equipped for writing
+robust, high-scale servers"][goal]. A [recent survey][survey] has shown that
+the biggest blocker to robust, high-scale servers is ergonomic usage of async
```

Comment:
Useful recent citation, though it seems rather basic. I'd like to investigate in more detail and the memory usage observation seems suspect. It does seem to support the preference for async though.

> Regarding the limited address space: x86_64 CPUs have AFAIK an addressable range of 128TB, which results in just 16k possible concurrent threads if each of them had the common 8MB stack space. So it's not like that's a thing of the past right?

I don't think so. x86\_64 has 48-bit physical addressing, and the usual kernel/userspace split uses the topmost bit to distinguish. So 2^47 / 8 MB is still millions of threads.